run.c

```diff
@@ -421,16 +421,49 @@ float random_f32() { // random float32 in [0,1)
     return (random_u32() >> 8) / 16777216.0f;
 }
 
-int sample(float* probabilities, int n) {
-    // sample index from probabilities, they must sum to 1
-    float r = random_f32();
-    float cdf = 0.0f;
+typedef struct {
+    float value;
+    int index;
+} IndexedFloat;
+
+int compare_indexed_float(const void* a, const void* b) {
+    IndexedFloat* indexed_a = (IndexedFloat*)a;
+    IndexedFloat* indexed_b = (IndexedFloat*)b;
+    // sorting in descending order
+    return (indexed_b->value > indexed_a->value) ? 1 : ((indexed_b->value < indexed_a->value) ? -1 : 0);
+}
+
+int sample_top_p(float* probs, int n, float topp) {
+    IndexedFloat probs_sort[n];
     for (int i = 0; i < n; i++) {
-        cdf += probabilities[i];
-        if (r < cdf) {
-            return i;
+        probs_sort[i].value = probs[i];
+        probs_sort[i].index = i;
+    }
+
+    qsort(probs_sort, n, sizeof(IndexedFloat), compare_indexed_float);
+
+    // Calculate cumulative probabilities
+    float cum_probs[n];
+    cum_probs[0] = probs_sort[0].value;
+    int p = 1;
+    for (; cum_probs[p-1]<=topp && p < n; p++) {
+        cum_probs[p] = cum_probs[p - 1] + probs_sort[p].value;
+    }
+
+    for (int i = 0; i < p; i++) {
+        probs_sort[i].value /= cum_probs[p-1];
```

Comment:
I think this rescaling can be avoided by scaling the sampled float `r` with `cum_probs[p-1]` instead.

Reply:
yeah, you are right, I have updated it.